src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs

```diff
@@ -0,0 +1,214 @@
+            if (indexOfFirstNonBase64 >= 0)
+            {
+                decodedLength = 0;
+                return false;
+            }
+
+            int length = base64Text.Length;
+            int paddingCount = 0;
+
+            // Check if there are chars that need to be ignored while determining the length
+            if (base64Text.IndexOfAny(CharsToIgnore) > -1)
+            {
+                fixed (char* srcChars = &MemoryMarshal.GetReference(base64Text))
+                {
+                    int numberOfCharsToIgnore = 0;
+                    char* src = srcChars;
+
+                    for (int i = 0; i < length; i++)
+                    {
+                        char charToValidate = *src++;
+                        if (IsCharToBeIgnored(charToValidate))
+                        {
+                            numberOfCharsToIgnore++;
+                        }
+                        else if (charToValidate == EncodingPad)
+                        {
+                            paddingCount++;
+                        }
+                    }
+
+                    length -= numberOfCharsToIgnore;
+                }
+            }
+            else if (length >= 4)
+            {
+                if (base64Text[length - 1] == EncodingPad)
+                {
+                    paddingCount++;
+                }
+                if (base64Text[length - 2] == EncodingPad)
+                {
+                    paddingCount++;
+                }
+            }
+
+            if (length % 4 != 0)
+            {
+                decodedLength = 0;
+                return false;
+            }
+
+            // Remove padding to get exact length
+            decodedLength = (length / 4 * 3) - paddingCount;
+            return true;
+        }
+
+        /// <summary>
+        /// Validates the span of UTF-8 encoded text represented as base64 into binary data.
+        /// </summary>
+        /// <param name="base64TextUtf8">The input span which contains UTF-8 encoded text in base64 that needs to be validated.</param>
+        /// <returns>true if <paramref name="base64TextUtf8"/> is decodable; otherwise, false.</returns>
+        public static bool IsValid(ReadOnlySpan<byte> base64TextUtf8) => IsValid(base64TextUtf8, out int _);
+
+        /// <summary>
+        /// Validates the span of UTF-8 encoded text represented as base64 into binary data.
+        /// </summary>
+        /// <param name="base64TextUtf8">The input span which contains UTF-8 encoded text in base64 that needs to be validated.</param>
+        /// <param name="decodedLength">The maximum length (in bytes) if you were to decode the base 64 encoded text <paramref name="base64TextUtf8"/> within a byte span.</param>
+        /// <returns>true if <paramref name="base64TextUtf8"/> is decodable; otherwise, false.</returns>
+        public static unsafe bool IsValid(ReadOnlySpan<byte> base64TextUtf8, out int decodedLength)
```

Comment:
The common case we want to optimize for is when there are no whitespace/invalid characters present.

The current implementation is doing two passes over the input. You can instead write it along the lines of
```c#
int indexOfPaddingInvalidOrWhitespace = base64Text.IndexOfAnyExcept(s_validBase64Chars);

int length;
if ((uint)indexOfPaddingInvalidOrWhitespace >= (uint)base64Text.Length)
{
    length = base64Text.Length;
}
else
{
    char c = base64Text[indexOfPaddingInvalidOrWhitespace];
    if (c != EncodingPad)
    {
        // ... count whitespace / check for invalid characters
    }

    // Check that the remaining chars are only '=' and whitespace
    // Check that the number of padding chars makes sense
}

decodedLength = length...

return (length & 3) == 0;
```